test: Add more tests for orphan tx handling

hebasto opened the pull request:
This PR adds test coverage for the following cases:
- erase orphan transactions when a peer is disconnected
- erase an orphan transaction when it is included in a new tip block
- erase an orphan transaction when it is conflicted with other transactions included in a new tip block

Found useful while working on #19374.

hebasto committed:
test: Add test for erase orphan tx from peer

hebasto committed:
test: Add test for erase orphan tx included by block

hebasto committed:
test: Add test for erase orphan tx conflicted by block

jonatack commented on test/functional/p2p_invalid_tx.py:
```diff
@@ -158,13 +157,71 @@ def run_test(self):
+        with node.assert_debug_log(['Erased 100 orphan tx from peer=24']):
+            self.reconnect_p2p(num_connections=1)
+
+        self.log.info('Test erase of orphan tx included by block')
```
What is being asserted in this subtest?

hebasto replied:
> What is being asserted in this subtest?

If a transaction from a new tip block is found in the orphan pool, it should be excluded from it.

jonatack replied:
Could add that info to the log message. Same with https://github.com/bitcoin/bitcoin/pull/19393#discussion_r446675398: state the action and expected result under test: "Test <action> causes <result>". Just my opinion, of course.

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19393#issuecomment-650802681).

jonatack commented on test/functional/p2p_invalid_tx.py:
```diff
@@ -158,13 +157,71 @@ def run_test(self):
+        tx_orphan_include_by_blockA.calc_sha256()
+
+        self.log.info('Send the tx_orphan_include_by_blockA ... ')
+        node.p2p.send_txs_and_test([tx_orphan_include_by_blockA], node, success=False)
```
This is checking the the txn is not accepted into the mempool; mind mentioning the expected outcome in the info log and why? ISTM this is the assertion for the subtest beginning at line 172? Perhaps just one log at 172 that describes the assertion/expected outcome/why, then one at the start of the next subtest that does the same.

hebasto replied:
Re-worded. It seems mention "orphan" transaction implies that it is not accepted into the mempool, no?

jonatack commented on test/functional/p2p_invalid_tx.py:
```diff
@@ -158,13 +157,71 @@ def run_test(self):
+        blockA.hashMerkleRoot = blockA.calc_merkle_root()
+        blockA.solve()
+
+        self.log.info('Send the blockA that includes tx_orphan_include_by_blockA ... ')
```
Idem, mind describing the expected outcome (accepted by the mempool) and why? This log seems more cryptic than reading the test code.

hebasto replied:
[Re-worded](https://github.com/bitcoin/bitcoin/pull/19393#issuecomment-650795757).

jonatack commented on test/functional/p2p_invalid_tx.py:
```diff
@@ -158,13 +157,71 @@ def run_test(self):
             node.p2p.send_txs_and_test([rejected_parent], node, success=False)
 
+        self.log.info('Test erase of orphan tx from peer')
+        with node.assert_debug_log(['Erased 100 orphan tx from peer=24']):
```
This test seems dependent on the context/test order. Is it possible to make it order-independent and to test for the change itself?

hebasto replied:
This variant, i.e., the previous test state reuse, just suggests the minimum diff.

jonatack reviewed (commented):
Concept ACK.

- Could say in the PR description what led you to see the need for this additional coverage.

- Testing debug log output is convenient but less robust; better to test for actual change, if possible.

- Readability nit: for variable names when the rest of the name is snakecased: `s/blockA/block_A` and `s/blockB/block_B/`

- It would be great for the logging to state clearly what the test is asserting, the expected outcome, and why. ISTM `info` logging can be good at the beginning of a subtest as well as just before the key assertion(s).

- If a subtest does not need the context from the previous ones, it can be good to separate it to an independent method.

hebasto commented:
@jonatack 
> Could say in the PR description what led you to see the need for this additional coverage.

The OP has been updated.

> Testing debug log output is convenient but less robust; better to test for actual change, if possible.

Agree. But we have neither RPC calls nor the public interface for the orphan transaction pool.

hebasto commented on test/functional/p2p_invalid_tx.py:
```diff
@@ -94,24 +93,24 @@ def run_test(self):
         tx_withhold = CTransaction()
         tx_withhold.vin.append(CTxIn(outpoint=COutPoint(block1.vtx[0].sha256, 0)))
-        tx_withhold.vout.append(CTxOut(nValue=50 * COIN - 12000, scriptPubKey=SCRIPT_PUB_KEY_OP_TRUE))
+        tx_withhold.vout = [CTxOut(nValue=25 * COIN  - 12000, scriptPubKey=SCRIPT_PUB_KEY_OP_TRUE)] * 2
```
hmm, why python linter did not catch it?

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19393#issuecomment-650795757).

hebasto commented:
Updated 10f7f75502bdcc9045c0623386571e41103b69fe -> a95a6d99b31798a8adb107930857bbe5c6e083df ([pr19393.01](https://github.com/hebasto/bitcoin/commits/pr19393.01) -> [pr19393.02](https://github.com/hebasto/bitcoin/commits/pr19393.02), [diff](https://github.com/hebasto/bitcoin/compare/pr19393.01..pr19393.02)):

- addressed @jonatack's comments

hebasto commented on test/functional/p2p_invalid_tx.py:
```diff
@@ -158,13 +157,71 @@ def run_test(self):
         with node.assert_debug_log(['not keeping orphan with rejected parents {}'.format(rejected_parent.hash)]):
             node.p2p.send_txs_and_test([rejected_parent], node, success=False)
 
+        self.log.info('Test erase of orphan tx from peer')
```
I'd rather leave log message as is, as working logging implies working erase a tx from the orphan pool.

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19393#issuecomment-650802681).

hebasto commented:
Updated a95a6d99b31798a8adb107930857bbe5c6e083df -> 3a310fbf1733e2ddaf1e6707251c409b287838de ([pr19393.02](https://github.com/hebasto/bitcoin/commits/pr19393.02) -> [pr19393.03](https://github.com/hebasto/bitcoin/commits/pr19393.03), [diff](https://github.com/hebasto/bitcoin/compare/pr19393.02..pr19393.03)):

- addressed @jonatack's [comment](https://github.com/bitcoin/bitcoin/pull/19393#discussion_r446676922):
> Could add that info to the log message. Same with [#19393 (comment)](https://github.com/bitcoin/bitcoin/pull/19393#discussion_r446675398): state the action and expected result under test: "Test (action) causes (result)". Just my opinion, of course.

hebasto commented:
Rebased 3a310fbf1733e2ddaf1e6707251c409b287838de -> 9979e8b96f475762c9ec4feb98bcd80c029b1913 ([pr19393.03](https://github.com/hebasto/bitcoin/commits/pr19393.03) -> [pr19393.04](https://github.com/hebasto/bitcoin/commits/pr19393.04)) due to the conflict with #19804.

hebasto commented:
Rebased 3a310fbf1733e2ddaf1e6707251c409b287838de -> 8e6c42797e0c5b30df34f464eb3b1cddc477da26 ([pr19393.03](https://github.com/hebasto/bitcoin/commits/pr19393.03) -> [pr19393.05](https://github.com/hebasto/bitcoin/commits/pr19393.05)) due to the conflict with #19804.

Emzy commented:
tACK
Tested and worked:
src/test/fuzz/addrman
src/test/fuzz/kitchen_sink
src/test/fuzz/merkleblock

NelsonGaldeman commented:
Concept ACK

- I have tested the code

leonardojobim commented:
Concept ACK. It expands the test cases that cause the removal of the orphan transactions.

I also agree it would be better to test against the actual changes, but as far I know there's no way in these tests to access  the orphan pool maintained in the `mapOrphanTransactionsdata` structure to perform this validation.

I ran successfully this test on Ubuntu 20.04.

leonardojobim reviewed (approved):
Concept ACK. It expands the test cases that cause the removal of the orphan transactions.

I also agree it would be better to test against the actual changes, but as far I know there's no way in these tests to access the orphan pool maintained in the `mapOrphanTransactionsdata` structure to perform this validation.

I ran successfully this test on Ubuntu 20.04.

aarmoa reviewed (approved):
Approach ACK.
To validate the new verification work as expected I did the following temporary change while running the test locally:
Line 168:
- If assertion is done for 99 number of orphans instead of 100, test fails.
- If assertion is done for node 23 instead of 24, test fails.

Line 190:
- Added a node.assert_debug_log() with the message "Erased 1 orphan tx included or conflicted by block" as unexpected, to ensure it is logged only when block_A is sent. The test runs without failures.

Line 219:
- Added a node.assert_debug_log() with the message "Erased 1 orphan tx included or conflicted by block" as unexpected, to ensure it is logged only when block_B is sent. The test runs without failures.

Tested in Linux Mint 20 64 bits.

practicalswift commented:
Concept ACK

ShubhamPalriwala commented on test/functional/p2p_invalid_tx.py:
```diff
@@ -93,24 +92,24 @@ def run_test(self):
         tx_withhold = CTransaction()
         tx_withhold.vin.append(CTxIn(outpoint=COutPoint(block1.vtx[0].sha256, 0)))
-        tx_withhold.vout.append(CTxOut(nValue=50 * COIN - 12000, scriptPubKey=SCRIPT_PUB_KEY_OP_TRUE))
+        tx_withhold.vout = [CTxOut(nValue=25 * COIN - 12000, scriptPubKey=SCRIPT_PUB_KEY_OP_TRUE)] * 2
```
nit: Having `append()` here would maintain consistency across the entire file

hebasto replied:
See the current code in the master branch https://github.com/bitcoin/bitcoin/blob/dd097c42df7378c9d0114c04c82b3c38c18e13dc/test/functional/p2p_invalid_tx.py#L102

ShubhamPalriwala commented on test/functional/p2p_invalid_tx.py:
```diff
@@ -93,24 +92,24 @@ def run_test(self):
         tx_orphan_1 = CTransaction()
         tx_orphan_1.vin.append(CTxIn(outpoint=COutPoint(tx_withhold.sha256, 0)))
-        tx_orphan_1.vout = [CTxOut(nValue=10 * COIN, scriptPubKey=SCRIPT_PUB_KEY_OP_TRUE)] * 3
+        tx_orphan_1.vout = [CTxOut(nValue=8 * COIN, scriptPubKey=SCRIPT_PUB_KEY_OP_TRUE)] * 3
```
nit: Having `append()` here would maintain consistency across the entire file

hebasto replied:
See the current code in the master branch https://github.com/bitcoin/bitcoin/blob/dd097c42df7378c9d0114c04c82b3c38c18e13dc/test/functional/p2p_invalid_tx.py#L102

ShubhamPalriwala commented on test/functional/p2p_invalid_tx.py:
```diff
@@ -157,13 +156,72 @@ def run_test(self):
+        self.log.info('Test that a transaction in the orphan pool is included in a new tip block causes erase this transaction from the orphan pool')
+        tx_withhold_until_block_A = CTransaction()
+        tx_withhold_until_block_A.vin.append(CTxIn(outpoint=COutPoint(tx_withhold.sha256, 1)))
+        tx_withhold_until_block_A.vout = [CTxOut(nValue=12 * COIN, scriptPubKey=SCRIPT_PUB_KEY_OP_TRUE)] * 2
```
nit: Having `append()` here would maintain consistency across the entire file

hebasto replied:
See the current code in the master branch https://github.com/bitcoin/bitcoin/blob/dd097c42df7378c9d0114c04c82b3c38c18e13dc/test/functional/p2p_invalid_tx.py#L102

ShubhamPalriwala reviewed (commented):
tACK 73278e3fda86f89ae70b8922f029e6b2d72d6bd6! 

Suggestion: the `log.info()` messages across the PR can be framed in a more clear and concise way.

I completely agree with @leonardojobim here!
> Concept ACK. It expands the test cases that cause the removal of the orphan transactions.
> 
> I also agree it would be better to test against the actual changes, but as far I know there's no way in these tests to access the orphan pool maintained in the `mapOrphanTransactionsdata` structure to perform this validation.
> 
> I ran successfully this test on Ubuntu 20.04.

Tested on Ubuntu 21.04 works fine

MarcoFalke commented:
> Tests don't pass for me on the first commit.

After a rebase, that is.

hebasto commented:
> Needs rebase.
> 
> Tests don't pass for me on the first commit.


Rebased 8e6c42797e0c5b30df34f464eb3b1cddc477da26 -> c0a5fceee9858afd24fe0bf655b7b30728e96e78 ([pr19393.05](https://github.com/hebasto/bitcoin/commits/pr19393.05) -> [pr19393.06](https://github.com/hebasto/bitcoin/commits/pr19393.06)).

pg156 reviewed (commented):
Reviewed to https://github.com/bitcoin/bitcoin/pull/19393/commits/c0a5fceee9858afd24fe0bf655b7b30728e96e78. Concept ACK. Agree due to the lack of RPC calls to inspect orphan pool, using `assert_debug_log` to match strings in log is a reasonable way to test.

Tested successfully by running `make check` and `test/functional/test_runner.py`.

If I understand correctly, https://github.com/bitcoin/bitcoin/pull/19393/commits/5c049780c8b310428cf72fb304bf0c1071742785 tests 100 orphan transactions sent by a peer to the node are removed after the peer is disconnected. (The peer initially sent 101 orphan transactions to the node. 1 orphan tx is removed because the default max orphan pool size `-maxorphantx` is 100.) This makes sense.

In https://github.com/bitcoin/bitcoin/commit/c0a5fceee9858afd24fe0bf655b7b30728e96e78, an orphan transaction is sent first to node, then a block with a conflicting transaction is sent to node, resulting in removing the orphan transaction. This is also reasonable.

Questions and suggestions are in below.

aureleoules commented:
tACK c0a5fceee9858afd24fe0bf655b7b30728e96e78 (`make check` and `test/functional/test_runner.py`).
Tested on NixOS 22.05 64 bits.  

I commented out locally [`src/txorphanage.cpp:189`](https://github.com/hebasto/bitcoin/blob/200627-test-orphan/src/txorphanage.cpp#L189) and the tests fail accordingly.  
Added tests seem to correctly test the behavior.  

nit: I would agree with [pg156](https://github.com/pg156), magic numbers should be changed to variables for increased readability and better comprehension.  

nit: I found it a bit confusing the use of `*_A` and `*_B` variables as I first thought the two tests were sequentially dependent. Would it make more sense to use unique variables that would get overridden by the second test, or separate both tests in different methods completely? I hope this is not too _nitty_ but it was my initial thought.

kouloumos commented on test/functional/p2p_invalid_tx.py:
```diff
@@ -157,13 +156,72 @@ def run_test(self):
+        tip = int(node.getbestblockhash(), 16)
+        height = node.getblockcount() + 1
+        block_A = create_block(tip, create_coinbase(height))
+        block_A.vtx.extend([tx_withhold, tx_withhold_until_block_A, tx_orphan_include_by_block_A])
```
When the grandchild was added it was missing inputs thus added to the orphan pool, but those orphans are not reevaluated at each newly connected block. 

Taking a look into the codebase, when a new orphan tx is added, there are a few indexes to the OrhanMap, [one of them is from the parents' COutPoint to the tx](https://github.com/hebasto/bitcoin/blob/c0a5fceee9858afd24fe0bf655b7b30728e96e78/src/txorphanage.cpp#L47-L49). On a newly connected block, [that index is used](https://github.com/hebasto/bitcoin/blob/c0a5fceee9858afd24fe0bf655b7b30728e96e78/src/txorphanage.cpp#L184) to check if the included transaction's parents have connection to any other txns from the orphan pool, and those are evicted. 

So in you example, when child (`tx_withhold_until_block_A`) is evaluated, eviction looks for txns in the pool also connected to the parent (`tx_withhold`) and not grandchildren (`tx_orphan_include_by_block_A`) connected to the child.

kouloumos commented:
ACK c0a5fceee9858afd24fe0bf655b7b30728e96e78 with a nit per https://github.com/bitcoin/bitcoin/pull/19393#discussion_r899156623.

The description of the flow in https://github.com/bitcoin/bitcoin/pull/19393#discussion_r900233637 might be useful for reviewers to understand the behavior tested on fa45bb21193ae0c220cfc224d5e3ea0e7f3ec988 & c0a5fceee9858afd24fe0bf655b7b30728e96e78.
